rasa/engine/validation.py

```diff
@@ -399,3 +401,66 @@ def _validate_parent_return_type(
             f"'{parent_return_type}' but type '{required_type}' "
             f"was expected by component '{node.uses.__name__}'."
         )
+
+
+def _validate_required_components(schema: GraphSchema,) -> None:
+    unmet_requirements: Dict[Text, Dict[Type, Set[Text]]] = dict()
+    for target_name in schema.target_names:
+        unmet_requirements[target_name], _ = _recursively_check_required_components(
```

Comment:
Maybe this would be simpler if we didn't worry about keying by target and just merged them together?